libs/ml/include/ml/optimisation/lazy_adam_optimiser.hpp

```diff
@@ -0,0 +1,371 @@
+                        gradient_tensor);
+  fetch::math::Multiply(cache_tensor, beta1_t_, cache_tensor);
+  fetch::math::Add(cache_tensor, gradient_tensor, cache_tensor);
+
+  // mt   = cache[i] / (1.0 - beta1_t_);
+  fetch::math::Divide(cache_tensor, (static_cast<DataType>(1.0) - beta1_t_), mt_tensor);
+
+  // momentum[i] = (beta2_t_ * momentum[i]) + ((1.0 - beta2_t_) *
+  // ((input_gradients[i]/batch_size)^2));
+  fetch::math::Divide(refs_tensor, static_cast<DataType>(batch_size), v_tensor);
+  fetch::math::Square(v_tensor, v_tensor);
+  fetch::math::Multiply(v_tensor, (static_cast<DataType>(1.0) - beta2_t_), v_tensor);
+  fetch::math::Multiply(momentum_tensor, beta2_t_, momentum_tensor);
+  fetch::math::Add(momentum_tensor, v_tensor, momentum_tensor);
+
+  // vt   = momentum[i] / (1.0 - beta2_t_);
+  fetch::math::Divide(momentum_tensor, (static_cast<DataType>(1.0) - beta2_t_), v_tensor);
+
+  // output_gradients[i] = -this->learning_rate_ * mt / (sqrt(vt) + epsilon_);
+  fetch::math::Sqrt(v_tensor, gradient_tensor);
+  fetch::math::Add(gradient_tensor, epsilon_, gradient_tensor);
+  fetch::math::Divide(mt_tensor, gradient_tensor, gradient_tensor);
+  fetch::math::Multiply(gradient_tensor, -this->learning_rate_, gradient_tensor);
+}
+
+template <class T>
+void LazyAdamOptimiser<T>::ApplyGradients(SizeType batch_size)
+{
+  // Do operation with gradient
+  auto cached_weight_it = cache_.begin();
+  auto momentum_it      = momentum_.begin();
+  auto mt_it            = mt_.begin();
+  auto vt_it            = vt_.begin();
+
+  auto gradient_it  = this->gradients_.begin();
+  auto trainable_it = this->graph_trainables_.begin();
+
+  // beta1_t=beta1^t and beta2_t=beta2^t, where t is number of epochs + 1
+  fetch::math::Pow(beta1_, static_cast<DataType>(this->epoch_ + 1), beta1_t_);
+  fetch::math::Pow(beta2_, static_cast<DataType>(this->epoch_ + 1), beta2_t_);
+
+  std::vector<SizeSet> rows;
+
+  while (gradient_it != this->gradients_.end())
+  {
+    // Skip frozen trainables
+    if ((*trainable_it)->GetFrozenState())
+    {
+      continue;
+    }
+
+    std::pair<TensorType const, SizeSet const> gradient_pair =
+        (*trainable_it)->GetSparseGradientsReferences();
+    rows.push_back(gradient_pair.second);
+
+    // Normal ApplyGradient
+    // if number_of_rows_to_update*sparsity_threshold_ > total_rows
+    if (rows.at(rows.size() - 1).empty() ||
+        (rows.at(rows.size() - 1).size() * sparsity_threshold_) > gradient_pair.first.shape().at(1))
+    {
+      ApplyLogic(batch_size, *gradient_it, *momentum_it, *mt_it, *vt_it, *cached_weight_it,
+                 (*trainable_it)->GetGradientsReferences());
+    }
+
+    else
+    {
+      // Sparse apply gradient
+      // if number_of_rows_to_update*sparsity_threshold_ <= total_rows
+
+      for (SizeType update_index : rows.at(rows.size() - 1))
+      {
+
+        auto       gradient_slice        = gradient_it->Slice(update_index, 1);
```

Comment:
I guess all this slicing and copying really eats into the benefits of the sparse update. Presumably there is no easy way to:
1. use views, and
2. reduce copying